File: src/haru_pack/overlay.py
from __future__ import annotations
import hashlib, struct
from pathlib import Path

MAGIC = b"HARUPACK"   # 8B start sentinel (matches launcher/overlay.nim FooterMagic)
TAIL  = b"KCAPURAH"   # 8B end sentinel   (FooterTail)
FOOTER_V1_SIZE = 8 + 2 + 2 + 8 + 8 + 32 + 8               # = 68  (payload only)
FOOTER_V2_SIZE = 8 + 2 + 2 + 8 + 8 + 32 + 8 + 8 + 32 + 8  # = 116 (payload + stub locator)
# v3 = v2 + a signature tail: an Ed25519 public key (32) + signature (64), placed OUTSIDE
# the signed bytes and BEFORE the TAIL, so the backward scan locates it by the same MAGIC.
FOOTER_V3_SIZE = FOOTER_V2_SIZE + 32 + 64                 # = 212 (v2 + pubkey + signature)
FOOTER_SIZE = FOOTER_V1_SIZE                              # back-compat alias (v1)

# The signed region of a v2/v3 footer: everything BETWEEN the MAGIC and the TAIL that is
# structural or a digest — formatVer, flags, payloadOff, payloadLen, payloadSha, stubOff,
# stubLen, stubSha. That is footer[8:108], 100 bytes. --self-signed signs exactly these
# bytes; the pubkey + signature that follow are deliberately NOT covered (a signature cannot
# authenticate itself). Because formatVer is inside the signed region, the signature also
# binds the format version.
_SIGNED_OFF, _SIGNED_END = 8, 108        # footer[8:108]
_PUBKEY_OFF, _PUBKEY_END = 108, 140      # footer[108:140]
_SIG_OFF, _SIG_END = 140, 204            # footer[140:204]

# Footer flag bits — MUST match launcher/overlay.nim FooterFlag* constants.
FOOTER_FLAG_ENCRYPTED = 1   # bit0: payload is an encrypted container (crypto.MAGIC)
FOOTER_FLAG_REMOTE = 2      # bit1: payload is fetched at runtime, not appended (Phase 3)

_FOOTER_SIZES = {1: FOOTER_V1_SIZE, 2: FOOTER_V2_SIZE, 3: FOOTER_V3_SIZE}


def _stub_footer(flags: int, off: int, paylen: int, paysha: bytes,
                 stub_off: int, stub_len: int, stub_sha: bytes, sign_key) -> tuple:
    """Build a stub-carrying footer (v2, or v3 when `sign_key` is given) and return
    (footer_bytes, extra_dict). The signed region is byte-identical between v2 and v3 apart
    from the format-version field, so the launcher's v2 reader and v3 reader share offsets.

    `sign_key` is a cryptography Ed25519 private key or None. When given, the footer is v3:
    the 100-byte signed region (with format_ver=3) is signed, and the public key + signature
    are appended after it, ahead of the TAIL — outside the signed bytes on purpose."""
    ver = 3 if sign_key is not None else 2
    signed = (struct.pack("<HHQQ", ver, flags, off, paylen) + paysha
              + struct.pack("<QQ", stub_off, stub_len) + stub_sha)
    assert len(signed) == _SIGNED_END - _SIGNED_OFF, len(signed)
    if sign_key is None:
        footer = MAGIC + signed + TAIL
        assert len(footer) == FOOTER_V2_SIZE, len(footer)
        return footer, {"format_ver": 2}
    pub = sign_key.public_key().public_bytes_raw()
    sig = sign_key.sign(signed)                     # Ed25519 is deterministic (RFC 8032)
    assert len(pub) == 32 and len(sig) == 64
    footer = MAGIC + signed + pub + sig + TAIL
    assert len(footer) == FOOTER_V3_SIZE, len(footer)
    return footer, {"format_ver": 3, "self_signed": True,
                    "pubkey": pub.hex(),
                    "pubkey_sha256": hashlib.sha256(pub).hexdigest()}


def attach(launcher: Path, payload: bytes, out: Path, flags: int = 0,
           stub_config: bytes | None = None, remote: bool = False, sign_key=None) -> dict:
    """Append `payload` (+ optional cleartext `stub_config`) + a versioned footer to the
    launcher PE/ELF. Run BEFORE signing: an Authenticode cert table appended afterward sits
    past the footer, which the launcher relocates by scanning backward for MAGIC.

    `stub_config=None` emits a v1 (68B) footer, byte-identical to the single-payload format
    every existing build and v1 test corpus loads. Passing bytes emits a v2 (116B) footer
    and lays the file out as [launcher][payload][stub-config][footer], the stub-config
    covered by the signature and located only by the footer (docs/adr/0003 §1).

    `remote=True` (Phase 3 remote-fetch, INV-REMOTE-01) records the payload DIGEST as the
    trust anchor but embeds NO payload bytes: the file is laid out [launcher][stub-config]
    [footer] with payload_len=0 and the remote flag bit set. The caller hosts `payload` at the
    stub-config's source_url; the launcher fetches it and verifies it against this digest, so
    the byte SOURCE is the only difference from an appended build. A remote build MUST carry a
    stub-config (it holds source_url).

    `sign_key` (a cryptography Ed25519 private key, or None) turns a stub-carrying build into a
    v3 --self-signed binary: the footer's structural + digest fields are signed and the public
    key + signature ride in a new footer tail (see `_stub_footer`). Signing REQUIRES a
    stub-config, because the signed region includes the stub locator; a v1 (no-stub) build
    cannot be self-signed. HONEST LIMIT: the public key is embedded in the same file, so this
    detects post-build edits by anyone who does not ALSO rewrite the key — it is not
    tamper-evidence unless the fingerprint is pinned out of band (docs/SIGNING.md)."""
    if sign_key is not None and stub_config is None:
        raise ValueError("--self-signed requires a stub-config: the signature covers the "
                         "stub locator, so a v1 (single-payload) footer cannot be signed")
    stub = Path(launcher).read_bytes()
    off = len(stub)
    paysha = hashlib.sha256(payload).digest()
    if remote:
        if stub_config is None:
            raise ValueError("remote build requires a stub_config (it carries source_url)")
        flags |= FOOTER_FLAG_REMOTE
        stub_off = off                     # stub-config right after the launcher; no payload between
        stub_sha = hashlib.sha256(stub_config).digest()
        footer, extra = _stub_footer(flags, off, 0, paysha, stub_off, len(stub_config),
                                     stub_sha, sign_key)
        Path(out).write_bytes(stub + stub_config + footer)
        info = {"remote": True, "payload_off": off, "payload_len": 0,
                "sha256": paysha.hex(), "stub_off": stub_off, "stub_len": len(stub_config),
                "stub_sha256": stub_sha.hex()}
        info.update(extra)
        return info
    if stub_config is None:
        footer = MAGIC + struct.pack("<HHQQ", 1, flags, off, len(payload)) + paysha + TAIL
        assert len(footer) == FOOTER_V1_SIZE, len(footer)
        Path(out).write_bytes(stub + payload + footer)
        return {"format_ver": 1, "payload_off": off, "payload_len": len(payload),
                "sha256": paysha.hex()}
    stub_off = off + len(payload)          # stub-config immediately after the payload
    stub_sha = hashlib.sha256(stub_config).digest()
    footer, extra = _stub_footer(flags, off, len(payload), paysha, stub_off, len(stub_config),
                                 stub_sha, sign_key)
    Path(out).write_bytes(stub + payload + stub_config + footer)
    info = {"payload_off": off, "payload_len": len(payload),
            "sha256": paysha.hex(), "stub_off": stub_off, "stub_len": len(stub_config),
            "stub_sha256": stub_sha.hex()}
    info.update(extra)
    return info


def verify(exe: Path) -> dict:
    """Locate the footer by backward scan (survives an appended signature), dispatch on
    format_ver, and confirm the payload (and, for v2, the stub-config) sha256 still match."""
    data = Path(exe).read_bytes()
    i = data.rfind(MAGIC)
    if i == -1:
        raise ValueError("no haru-pack footer found")
    ver = struct.unpack_from("<H", data, i + 8)[0]
    size = _FOOTER_SIZES.get(ver)
    if size is None:
        raise ValueError(f"unsupported footer version {ver}")
    footer = data[i:i + size]
    if len(footer) != size or footer[size - 8:size] != TAIL:
        raise ValueError("footer tail sentinel mismatch")
    _, flags, off, ln = struct.unpack("<HHQQ", footer[8:28])
    paysha = footer[28:60]
    payload = data[off:off + ln]
    out = {"footer_at": i, "format_ver": ver, "flags": flags,
           "payload_off": off, "payload_len": ln,
           "sha_ok": hashlib.sha256(payload).digest() == paysha,
           "bytes_after_footer": len(data) - (i + size)}
    if ver in (2, 3):
        stub_off, stub_len = struct.unpack("<QQ", footer[60:76])
        stub_sha = footer[76:108]
        stub = data[stub_off:stub_off + stub_len]
        out.update({"stub_off": stub_off, "stub_len": stub_len,
                    "stub_ok": hashlib.sha256(stub).digest() == stub_sha,
                    "stub_config": stub.decode("utf-8", "replace")})
    if ver == 3:
        # v3 --self-signed: verify the Ed25519 signature over the footer's signed region
        # against the EMBEDDED public key. sig_ok=True means "these structural+digest fields
        # were signed by whoever's key is embedded here" — NOT that the key is trusted. That
        # is the documented honest limit (INV-SIGN-01); pin pubkey_sha256 out of band for it
        # to mean anything about provenance.
        signed = footer[_SIGNED_OFF:_SIGNED_END]
        pub = footer[_PUBKEY_OFF:_PUBKEY_END]
        sig = footer[_SIG_OFF:_SIG_END]
        out.update({"self_signed": True, "pubkey": pub.hex(),
                    "pubkey_sha256": hashlib.sha256(pub).hexdigest(),
                    "sig_ok": _ed25519_verify(pub, sig, signed)})
    return out


def _ed25519_verify(pub: bytes, sig: bytes, msg: bytes) -> bool:
    """True iff `sig` is a valid Ed25519 signature over `msg` under raw public key `pub`.
    Imported lazily so a caller that only builds v1/v2 binaries pays nothing for it."""
    from cryptography.exceptions import InvalidSignature
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
    try:
        Ed25519PublicKey.from_public_bytes(pub).verify(sig, msg)
        return True
    except (InvalidSignature, ValueError):
        return False

File: src/haru_pack/launcher/overlay.nim
## haru-pack overlay: locate an appended payload (and, on a v2 build, a cleartext
## stub-config section) inside our own signed PE. The footer is fixed-size and located by
## scanning BACKWARD for the start magic, so an Authenticode cert table appended at EOF
## (post-signing) does not hide it (INV-LAUNCH-01/05).
##
## Versioned footer (docs/adr/0003-stub-config-and-canary.md §1): the reader dispatches on
## `format_ver`, so a v1 (68B, single-payload) binary keeps loading byte-for-byte while a
## v2 (116B) binary additionally carries the stub-config locator. The shared prefix [0:60]
## is identical across versions; TAIL is always the final 8 bytes.
import std/[os, streams]

const
  FooterMagic* = "HARUPACK"   # 8B start sentinel
  FooterTail*  = "KCAPURAH"   # 8B end sentinel
  FooterV1Size* = 8 + 2 + 2 + 8 + 8 + 32 + 8            # = 68  (payload only)
  FooterV2Size* = 8 + 2 + 2 + 8 + 8 + 32 + 8 + 8 + 32 + 8  # = 116 (payload + stub locator)
  FooterV3Size* = FooterV2Size + 32 + 64                # = 212 (v2 + Ed25519 pubkey + sig)
  SignedRegionLen* = 100      # footer[8:108]: formatVer..stubSha, the --self-signed bytes
  ScanWindow*  = 256 * 1024   # bytes from EOF to search
  FooterFlagEncrypted* = 1'u16 ## bit0: payload is an encrypted container (matches crypto.py)
  FooterFlagRemote*    = 2'u16 ## bit1: payload is FETCHED at runtime, not appended (Phase 3,
                               ## INV-REMOTE-01). payloadLen is 0 and no payload bytes are
                               ## embedded; payloadSha remains the trust anchor for the bytes
                               ## fetched from the SOURCE_URL knob.

type Footer* = object
  formatVer*: uint16
  flags*: uint16
  payloadOff*: uint64
  payloadLen*: uint64
  payloadSha*: array[32, byte]
  hasStub*: bool              # true for a v2 or v3 footer
  stubOff*: uint64
  stubLen*: uint64
  stubSha*: array[32, byte]
  hasSig*: bool               # true only for a v3 (--self-signed) footer
  signed*: array[100, byte]   # footer[8:108], the bytes the signature covers (v2/v3)
  pubKey*: array[32, byte]    # v3: embedded Ed25519 public key (OUTSIDE the signed region)
  sig*: array[64, byte]       # v3: Ed25519 signature over `signed`

proc rdU16(b: openArray[byte], o: int): uint16 =
  uint16(b[o]) or (uint16(b[o+1]) shl 8)
proc rdU64(b: openArray[byte], o: int): uint64 =
  var v: uint64 = 0
  for i in 0..7: v = v or (uint64(b[o+i]) shl (8*i))
  v

proc footerSizeFor(ver: uint16): int =
  ## Byte size of a footer at this version, or 0 for a version this launcher does not
  ## understand — an unknown version is treated as a false-positive MAGIC, never parsed.
  case ver
  of 1'u16: FooterV1Size
  of 2'u16: FooterV2Size
  of 3'u16: FooterV3Size
  else: 0

proc findFooter*(exePath: string): (bool, Footer, int) =
  ## returns (found, footer, absoluteFooterOffset). Dispatches on format_ver: the real
  ## footer's MAGIC is the highest-indexed one, so a MAGIC that happens to occur inside the
  ## payload or stub-config has a lower index and is not reached first.
  var f = newFileStream(exePath, fmRead)
  defer: f.close()
  let size = getFileSize(exePath).int
  let winStart = max(0, size - ScanWindow)
  f.setPosition(winStart)
  let win = f.readStr(size - winStart)
  # start where an 8-byte MAGIC plus its 2-byte format_ver can still be read in-window
  var i = win.len - 10
  while i >= 0:
    if win[i ..< i+8] == FooterMagic:
      let ver = uint16(byte(win[i+8])) or (uint16(byte(win[i+9])) shl 8)
      let fsize = footerSizeFor(ver)
      # An unknown version here is NOT our footer (fail closed by scanning on); a known
      # version must fit in-window AND carry TAIL at its own end, or it is a false positive.
      if fsize > 0 and i + fsize <= win.len and
         win[i+fsize-8 ..< i+fsize] == FooterTail:
        var raw = newSeq[byte](fsize)
        for k in 0 ..< fsize: raw[k] = byte(win[i+k])
        var ft: Footer
        ft.formatVer  = rdU16(raw, 8)
        ft.flags      = rdU16(raw, 10)
        ft.payloadOff = rdU64(raw, 12)
        ft.payloadLen = rdU64(raw, 20)
        for k in 0..31: ft.payloadSha[k] = raw[28+k]
        if ver == 2'u16 or ver == 3'u16:
          ft.hasStub = true
          ft.stubOff = rdU64(raw, 60)
          ft.stubLen = rdU64(raw, 68)
          for k in 0..31: ft.stubSha[k] = raw[76+k]
          # Capture the signed region (footer[8:108]) for v2 and v3 alike; only v3 carries a
          # signature over it, but the bytes are identical, so it is cheap to always keep.
          for k in 0 ..< SignedRegionLen: ft.signed[k] = raw[8+k]
        if ver == 3'u16:
          # v3 --self-signed: the Ed25519 public key and signature ride AFTER the signed
          # region and before TAIL (raw[108:140] pubkey, raw[140:204] sig). They are outside
          # `signed` on purpose — a signature cannot authenticate itself.
          ft.hasSig = true
          for k in 0..31: ft.pubKey[k] = raw[108+k]
          for k in 0..63: ft.sig[k] = raw[140+k]
        return (true, ft, winStart + i)
    dec i
  var empty: Footer
  return (false, empty, -1)

proc footerFault*(ft: Footer, fileSize, footerAt: int): string =
  ## "" when the footer's declared extents are representable and actually lie inside the
  ## file, ahead of the footer. Anything else is a diagnostic string.
  ##
  ## payloadOff/payloadLen (and, on v2, stubOff/stubLen) come off disk, i.e. from whoever
  ## last wrote to the binary. Casting them straight to `int` and handing them to
  ## setPosition/readStr asks the runtime to allocate an attacker-chosen length; -d:release
  ## keeps Nim's bounds and range checks so the result is a crash or an OOM rather than
  ## memory corruption, but a launcher that aborts on a hostile footer with a raw Nim error
  ## is still doing input validation by accident. Do it on purpose (INV-LAUNCH-05).
  if fileSize <= 0: return "cannot determine the size of the executable"
  if footerAt < 0 or footerAt > fileSize: return "footer offset is outside the file"
  let fs = uint64(fileSize)
  # Phase 3 remote-fetch (INV-REMOTE-01): a remote build embeds NO payload bytes, so the
  # payload extent is legitimately empty (payloadLen == 0). Skip the payload-extent checks
  # for such a footer; readPayload is never called on it (main.launch fetches instead), and
  # the stub-config — which is mandatory for a remote build (it carries source_url) — still
  # gets the full validation below. An APPENDED build (the default) is unchanged.
  let remote = (ft.flags and FooterFlagRemote) != 0'u16
  if remote:
    # A remote footer embeds no payload: its extent MUST be empty. readPayload is never called
    # on this path, but pin the fields to their only valid values so no future reader inherits
    # an unvalidated, attacker-writable offset/length from a remote-flagged footer.
    if ft.payloadLen != 0'u64:
      return "remote footer declares a non-zero payload length"
    if ft.payloadOff > fs:
      return "remote footer payload offset lies past the end of the file"
  if not remote:
    if ft.payloadLen == 0'u64: return "footer declares a zero-length payload"
    # Compare each term against the file size BEFORE adding them, so the sum cannot wrap.
    if ft.payloadOff > fs: return "payload offset lies past the end of the file"
    if ft.payloadLen > fs: return "payload length exceeds the size of the file"
    if ft.payloadOff + ft.payloadLen > fs:
      return "payload extends past the end of the file"
    if ft.payloadOff + ft.payloadLen > uint64(footerAt):
      return "payload overlaps its own footer"
  # v2 only: the stub-config locator gets the same up-front validation (INV-LAUNCH-08).
  if ft.hasStub:
    if ft.stubLen == 0'u64: return "footer declares a zero-length stub-config"
    if ft.stubOff > fs: return "stub-config offset lies past the end of the file"
    if ft.stubLen > fs: return "stub-config length exceeds the size of the file"
    if ft.stubOff + ft.stubLen > fs:
      return "stub-config extends past the end of the file"
    if ft.payloadOff + ft.payloadLen > ft.stubOff:
      return "payload overlaps the stub-config"
    if ft.stubOff + ft.stubLen > uint64(footerAt):
      return "stub-config overlaps its own footer"
  return ""

proc readPayload*(exePath: string, ft: Footer, footerAt: int): string =
  ## Read the raw payload bytes described by the footer, refusing a footer whose
  ## extent does not fit inside this file (see footerFault).
  let fault = footerFault(ft, getFileSize(exePath).int, footerAt)
  if fault.len > 0: raise newException(ValueError, "corrupt payload footer: " & fault)
  var f = newFileStream(exePath, fmRead)
  defer: f.close()
  f.setPosition(int(ft.payloadOff))
  result = f.readStr(int(ft.payloadLen))
  if result.len != int(ft.payloadLen):
    raise newException(ValueError, "payload is shorter than its footer declares")

proc readStub*(exePath: string, ft: Footer, footerAt: int): string =
  ## Read the raw stub-config bytes described by a v2 footer. Its extent is validated by
  ## the same footerFault the payload uses (INV-LAUNCH-05/08 discipline); the digest over
  ## these bytes is checked by the caller before parse (INV-STUB-01).
  let fault = footerFault(ft, getFileSize(exePath).int, footerAt)
  if fault.len > 0: raise newException(ValueError, "corrupt stub-config footer: " & fault)
  var f = newFileStream(exePath, fmRead)
  defer: f.close()
  f.setPosition(int(ft.stubOff))
  result = f.readStr(int(ft.stubLen))
  if result.len != int(ft.stubLen):
    raise newException(ValueError, "stub-config is shorter than its footer declares")
